src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -1251,34 +1275,47 @@ static ShouldWatch fromString(String s) throws EvalException {
 
   protected void maybeWatch(StarlarkPath starlarkPath, ShouldWatch shouldWatch)
       throws EvalException, RepositoryFunctionException, InterruptedException {
-    if (shouldWatch == ShouldWatch.NO) {
+    RepoCacheFriendlyPath repoCacheFriendlyPath =
+        toRepoCacheFriendlyPath(starlarkPath.getPath(), shouldWatch);
+    if (repoCacheFriendlyPath == null) {
       return;
     }
-    Path path = starlarkPath.getPath();
-    Pair<RepoRecordedInput.File, RootedPath> pair = toRootedPath(path);
-    if (path.startsWith(workingDirectory)) {
-      // The file is under the working directory. Don't watch it, as it would cause a dependency
-      // cycle.
-      if (shouldWatch == ShouldWatch.AUTO) {
-        return;
-      }
-      throw Starlark.errorf("attempted to watch file under working directory");
-    }
-    if (!allowWatchingFilesOutsideWorkspace && pair.first instanceof FileOutsideWorkspace) {
-      if (shouldWatch == ShouldWatch.AUTO) {
-        return;
-      }
-      throw Starlark.errorf(
-          "attempted to watch file outside workspace, but it's prohibited in the current context");
+    RootedPath rootedPath = repoCacheFriendlyPath.getRootedPath(env, directories);
+    if (rootedPath == null) {
+      throw new NeedsSkyframeRestartException();
     }
     try {
       FileValue fileValue =
-          (FileValue) env.getValueOrThrow(FileValue.key(pair.second), IOException.class);
+          (FileValue) env.getValueOrThrow(FileValue.key(rootedPath), IOException.class);
       if (fileValue == null) {
         throw new NeedsSkyframeRestartException();
       }
 
-      recordedFileInputs.put(pair.first, RepoRecordedInput.File.fileValueToMarkerValue(fileValue));
+      recordedFileInputs.put(
+          new RepoRecordedInput.File(repoCacheFriendlyPath),
+          RepoRecordedInput.File.fileValueToMarkerValue(fileValue));
+    } catch (IOException e) {
+      throw new RepositoryFunctionException(e, Transience.TRANSIENT);
+    }
+  }
+
+  protected void maybeWatchDirents(Path path, ShouldWatch shouldWatch)
+      throws EvalException, RepositoryFunctionException, InterruptedException {
+    RepoCacheFriendlyPath repoCacheFriendlyPath = toRepoCacheFriendlyPath(path, shouldWatch);
+    if (repoCacheFriendlyPath == null) {
+      return;
+    }
+    RootedPath rootedPath = repoCacheFriendlyPath.getRootedPath(env, directories);
+    if (rootedPath == null) {
```

Comment:
nit: use `env.valuesMissing()` instead? (not that big a deal, but the nullness check a few lines above triggers in a very different case so it's slightly confusing)